File: Algorithm/javascript/BAEKJOON/CLASS 2/4949.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on("line", (line) => {
  if (!line) {
    rl.close();
  } else {
    input.push(line);
  }
}).on("close", () => {
  solution(input);
  process.exit();
});

function solution(input) {
  for (let i = 0; i < input.length - 1; i++) {
    let string = input[i];
    let bracketArray = [];
    for (character of string) {
      if (character == "(" || character == "[") {
        bracketArray.push(character);
      } else if (character == ")") {
        if (
          bracketArray.length != 0 &&
          bracketArray[bracketArray.length - 1] == "("
        ) {
          bracketArray.pop();
        } else {
          bracketArray.push(character);
        }
      } else if (character == "]") {
        if (
          bracketArray.length != 0 &&
          bracketArray[bracketArray.length - 1] == "["
        ) {
          bracketArray.pop();
        } else {
          bracketArray.push(character);
        }
      }
    }
    if (bracketArray.length == 0) {
      console.log("yes");
    } else {
      console.log("no");
    }
  }
}

File: Algorithm/python/BAEKJOON/CLASS 2/4949.py
# TRY 1
'''
- string의 종료가 "."으로 결정된다는 것의 표현이 어려웠음
- bracket_list의 조건(길이가 0이 아니고, character가 닫히는 괄호이다)을 충족시키는 코드의 표현 어려움
'''
while True:
    string = input()
    if string == ".":
        break
    bracket_list = []
    for character in string:
        if character == "(" or character == "[":
            bracket_list.append(character)
        elif character == ")":
            if len(bracket_list) != 0 and bracket_list[-1] == "(":
                bracket_list.pop()
            else:
                bracket_list.append(character)
        elif character == "]":
            if len(bracket_list) != 0 and bracket_list[-1] == "[" :
                bracket_list.pop()
            else:
                bracket_list.append(character)
    if len(bracket_list) == 0:
        print("yes")
    else:
        print("no")
